translated/tech/20190715-Passing-callbacks-and-pointers-to-Cgo.md

```diff
@@ -0,0 +1,253 @@
+# 传递指针和回调函数到Cgo
+
+`Cgo`允许Go程序调用C库或其他暴露了C接口的库。正是如此，这也成为 Go 程序员工具箱的重要组成部分。
+
+使用`Cgo`可能会比较棘手，特别是在 Go 和 C 代码中传递指针和回调函数时。
+这篇文章讨论了一个端到端当例子，包含了如下几方面：
+* `Cgo`的基本使用，包括链接一个传统的C库到Go二进制文件中。
+* 从Go语言中传递struct到C语言中。
+* 传递Go函数到C程序中，并安排C程序在随后调用它们。
+* 安全到传递任意的Go数据到C代码中，这些C代码后续会回传这些数据到它所调用的Go回调中。
+
+本文并不是一个`Cgo`的使用教程-在阅读前，需要你对它对简单使用案例有所熟悉。
+在本文最后列了一些有用的`Cgo`使用教程和相关的文章。这个案例的全部源代码详见[Github](https://github.com/eliben/code-for-blog/tree/master/2019/cgo-callback)。
+
+## 问题所在-一个C库调用多个Go回调程序
+
+如下是一个虚构的C库的头文件，该库是通过一些数据和基于事件的回调来运行。
+
+````cgo
+typedef void (*StartCallbackFn)(void* user_data, int i);
+typedef void (*EndCallbackFn)(void* user_data, int a, int b);
+
+typedef struct {
+  StartCallbackFn start;
+  EndCallbackFn end;
+} Callbacks;
+// Processes the file and invokes callbacks from cbs on events found in the
+// file, each with its own relevant data. user_data is passed through to the
+// callbacks.
+void traverse(char* filename, Callbacks cbs, void* user_data);
+````
+
+回调标签是由几个重要的模式组成，所展示的这些模式在现实中也同样普遍:
+
+* 每一个回调拥有自己的类型签名，这里为了简便，我们使用`int`类型的参数，这个参数可以是其他任何类型。
+* 当只有较小数量的回调被调用时，它们可能作为独立的参数被传递到traverse中；然而，回调到数量会非常大（比如说，超过三个），
+然后几乎总是有一个汇集它们的结构体被传递，然后几乎总是有一个收集它们的结构被传递。
+通常，允许用户将某些回调设置为NULL传递到库中，这些特定事件对这种回调不感兴趣，并且不应为此调用任何用户代码。
+* 每个回调通过对traverse的调用都获得一个模糊的user_data指针。它用于区分互不相同的遍历，并传递用户特定的状态。
+遍历通常会传递user_data甚至不尝试访问它； 由于它是`void *`，因此它对于库是完全模糊的，
+并且用户代码会将其强制转换为回调中的某些具体类型。
+
+我们对traverse的实现仅是一个简单的模拟：
+
+````cgo
+void traverse(char* filename, Callbacks cbs, void* user_data) {
+  // 模拟某些遍历，调用start回调，之后调用end回调
+  // callback, if they are defined.
+  if (cbs.start != NULL) {
+    cbs.start(user_data, 100);
+  }
+  if (cbs.end != NULL) {
+    cbs.end(user_data, 2, 3);
+  }
+}
+````
+
+我们的任务是包装这个库，在Go代码中进行使用。我们想要在遍历中调用Go回调，不用再写任何多余的C代码。
+
+## Go 接口
+
+让我们从构思在Go代码中我们接口的样式开始，如下是一个方式：
+
+````
+type Visitor interface {
+  Start(int)
+  End(int, int)
+}
+func GoTraverse(filename string, v Visitor) {
+  // ... 实现
+}
+````
+
+文章的其余部分显示了使用此方法的完整实现。但是，它有一些缺点：
+* 当我们需要提供大量的回调时，如果我们仅对几个回调感兴趣，编写Visitor的实现可能会很乏味。
+可以通过提供一种结构来实现带有某些默认值（例如，无操作）的完整接口来减轻这种情况，然后用户结构可以嵌入此默认结构，而不必实现每个方法。
```

Comment:
This can be mitigated by providing a struct to implement the complete interface with some defaults (say, no-ops) and user structs can then embed this default struct and not have to implement every single method.
struct 这里的 struct 要翻译成结构体，或者不翻译
with some default 这里是默认操作，不应该是默认值
user structs can then embed   这里的嵌入应该是指匿名继承